server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -224,14 +241,17 @@ public final void write(final T state, final Path... locations) throws WriteStat
             copyStateToExtraLocations(directories, tmpFileName);
             performRenames(tmpFileName, fileName, directories);
             performStateDirectoriesFsync(directories);
+        } catch (WriteStateException e) {
+            cleanupOldFiles(oldGenerationId, locations);
```

Comment:
I think this is dangerous and can lead to data loss. Assume that you've successfully written a cluster state that contains an index "test" with state file of generation 1. Then you try to write an updated cluster state for index with state file generation 2. Writing the state file for the index is successful, but there's a failure later to write the state file for another index. Now the node crashes or the clean-up logic fails. When you then handle the next cluster state, you will try to write generation 3, which, if it fails, will clean up everything except generation 2.

I think that within MetaDataStateFormat, you can only clean up files within the write method that you know that you have written. In particular, you can't assume (given the manifest-based approach) that the highest generation you see here is the file to keep around (as the manifest might not be pointing to the one with the highest generation).

This clean-up logic will have to be handled at a higher level (GatewayMetaState)

Reply:
This is a really good catch! Thank you! The reason why I've implemented it this way, because I DO know generations of global metadata and index metadata from the manifest file, but I DON'T know the previous generation of the manifest file to perform cleanup of newly created manifest state file if the manifest write fails. I really would like to avoid returning manifest generation from loadFullState, because it already returns Tuple. Adding generation field to Manifest itself also feels wrong. So instead I've done the work in 2 commits:
1. MetaDataStateFormat.write should not perform cleanup at all, writeAndCleanup should alwayshttps://github.com/elastic/elasticsearch/pull/35049/commits/326be60539111cb922c973727dd890ea09747f48
2. Use writeManifestAndCleanup, reorder rollbackActions before write https://github.com/elastic/elasticsearch/pull/35049/commits/982e853dfef534e07a35c56478c20019c2161d01
From my point of view, this should work.